src/rust/sqs-executor/src/retriever/event_retriever.rs

```diff
@@ -0,0 +1,231 @@
+use std::io::{Read, Stdout};
+use std::marker::PhantomData;
+use std::time::Duration;
+
+use async_trait::async_trait;
+use rusoto_s3::{GetObjectError, GetObjectRequest, S3};
+use rusoto_sqs::Message as SqsMessage;
+
+use tokio::prelude::*;
+use tracing::{debug, error, info};
+
+use crate::errors::{CheckedError, Recoverable};
+use crate::event_decoder::PayloadDecoder;
+use futures::FutureExt;
+use grapl_observe::metric_reporter::{tag, HistogramUnit, MetricReporter};
+use grapl_observe::timers::{time_it, TimedFutureExt};
+use rusoto_core::RusotoError;
+use std::collections::HashMap;
+use tokio::time::Elapsed;
+
+#[async_trait]
+pub trait PayloadRetriever<T> {
+    type Message;
+    type Error: CheckedError;
+    async fn retrieve_event(&mut self, msg: &Self::Message) -> Result<Option<T>, Self::Error>;
+}
+
+pub struct S3PayloadRetriever<S, SInit, D, E, DecoderErrorT>
```

Comment:
this struct also exists in the similarly named file s3_event_retriever; seems like some deduping is needed